nixos/modules/services/databases/postgresql.md

```diff
@@ -42,68 +42,51 @@ services.postgresql.dataDir = "/data/postgresql";
 ## Upgrading {#module-services-postgres-upgrading}
 
 ::: {.note}
-The steps below demonstrate how to upgrade from an older version to `pkgs.postgresql_13`.
+The steps below demonstrate how to upgrade from an older version to `pkgs.postgresql_15`.
 These instructions are also applicable to other versions.
 :::
 
-Major PostgreSQL upgrades require a downtime and a few imperative steps to be called. This is the case because
-each major version has some internal changes in the databases' state during major releases. Because of that,
+Major PostgreSQL upgrades require a downtime. This is the case because each major version
+has some internal changes in the databases' state during major releases. Because of that,
 NixOS places the state into {file}`/var/lib/postgresql/&lt;version&gt;` where each `version`
 can be obtained like this:
 ```
 $ nix-instantiate --eval -A postgresql_13.psqlSchema
 "13"
 ```
-For an upgrade, a script like this can be used to simplify the process:
+For an upgrade, the `services.postgresql.upgradeFrom` option can be used:
 ```
 { config, pkgs, ... }:
 {
-  environment.systemPackages = [
-    (let
-      # XXX specify the postgresql package you'd like to upgrade to.
-      # Do not forget to list the extensions you need.
-      newPostgres = pkgs.postgresql_13.withPackages (pp: [
-        # pp.plv8
-      ]);
-    in pkgs.writeScriptBin "upgrade-pg-cluster" ''
-      set -eux
-      # XXX it's perhaps advisable to stop all services that depend on postgresql
-      systemctl stop postgresql
-
-      export NEWDATA="/var/lib/postgresql/${newPostgres.psqlSchema}"
-
-      export NEWBIN="${newPostgres}/bin"
-
-      export OLDDATA="${config.services.postgresql.dataDir}"
-      export OLDBIN="${config.services.postgresql.package}/bin"
-
-      install -d -m 0700 -o postgres -g postgres "$NEWDATA"
-      cd "$NEWDATA"
-      sudo -u postgres $NEWBIN/initdb -D "$NEWDATA"
-
-      sudo -u postgres $NEWBIN/pg_upgrade \
-        --old-datadir "$OLDDATA" --new-datadir "$NEWDATA" \
-        --old-bindir $OLDBIN --new-bindir $NEWBIN \
-        "$@"
-    '')
-  ];
-}
-```
+  services.postgresql = {
+    package = pkgs.postgresql_15;
+
+    upgradeFrom = {
+      package = pkgs.postgresql_13;
+
+      # You may specify a different location of the old data directory
```

Comment:
This is very unfortunate. We shouldn't have to specify any of the old state.
The old version of the package can be tracked with an indirect garbage collection root symlink.
The old location needs a small change. Currently it's the user's responsibility to make the postgres major version part of the data path. This makes the old state hard to find. I think we should let the user pick the path _up to_ the versioned directory instead. That way we have a single location where we can track the current package and data directory location.

Let's define this directory to be the "database directory" (as opposed to just "data directory").
```nix
options.services.postgresql.databaseDir = mkOption {
  description = ''
    Version-independent location of the database data directories.
    Unlike `dataDir`, this value must not contain the postgresql version.
  '';
  default = "/var/lib/postgresql";
};
```

Example layout (assuming default databaseDir)
```
/var/lib/postgresql/14
/var/lib/postgresql/14/nix-postgresql -> /nix/store/...
/var/lib/postgresql/14/*           # postgresql data
/var/lib/postgresql/15
/var/lib/postgresql/15/nix-postgresql -> /nix/store/...
/var/lib/postgresql/15/*           # postgresql data
/var/lib/postgresql/current -> 14
/var/lib/postgresql/new -> 15      # only during upgrade
```

As part of server startup, a script checks that `current` points to the intended major version.
If so, starts the server and updates the `nix-postgresql` symlink
If not, performs the migration using
 - `OLD` variables are filled based on `current` symlink
 - `NEW` variables are set as desired based on the NixOS configuration
 - `new` symlink is set to the new major version
 - perform migration
 - add `nix-postgresql` symlink in `new`
 - validate and/or start (could probably start the server using the absolute location e.g. /var/lib/postgresql/15 before switching the symlink)
 - `mv new current`
 - migration done

By organizing the migration in this way, you don't have to configure the old version, which is great because that avoids mistakes and config churn, and is overall easier to use.